src/ui/public/vis/editors/default/default.js

```diff
@@ -127,6 +137,32 @@ const defaultEditor = function ($rootScope, $compile) {
             catch (e) {} // eslint-disable-line no-empty
           }, true);
 
+          $scope.$watch(() => {
```

Comment:
I think we can make this a little easier to understand with unabbreviated variable names, variables to reference the object properties, and early exits:

```js
          $scope.$watch(() => {
            return $scope.vis.getCurrentState(false);
          }, (newState) => {
            // figure out what changed and copy into editor state
            const updateEditorStateWithChanges = (newState, oldState, editorState) => {
              for (const prop in newState) {
                const newStateValue = newState[prop];
                const oldStateValue = oldState[prop];
                const editorStateValue = editorState[prop];

                if (typeof newStateValue === 'object') {
                  if (editorStateValue) {
                    // Keep traversing.
                    return updateEditorStateWithChanges(newStateValue, oldStateValue, editorStateValue);
                  }
                
                  const newStateValueCopy = _.cloneDeep(newStateValue);
                  editorState[prop] = newStateValueCopy;
                  oldState[prop] = newStateValueCopy;
                  return;
                }

                if (newStateValue !== oldStateValue) {
                  oldState[prop] = newStateValue;
                  editorState[prop] = newStateValue;
                  lockDirty = true;
                }
              }
            };

            updateEditorStateWithChanges(newState, $scope.oldState, $scope.state);
          }, true);
```